### PR TITLE
Implement drag and drop reordering in General files list

### PR DESCRIPTION
**Pull request by Siedlerchr**
Add drag and drop related methods to ViewModelFactory
Make LInkedFile serializeable
Fixes https://github.com/JabRef/jabref/issues/3165#issuecomment-326269715

- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [ ] Screenshots added (for bigger UI changes)
- [x] Manually tested changed features in running JabRef
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)
- [ ] If you changed the localization: Did you run `gradle localizationUpdate`?

### CODE REVIEW

**Comment by Siedlerchr** [comment]
@lenhard Hm, that's odd. I will try to look at it. But I think it could be possible that there is no database changed event fired

**Comment by Siedlerchr** [comment]
@lenhard  The file ordering is now also visible in the bibtex data.

File: src/main/java/org/jabref/model/entry/LinkedFile.java [hunk 1]
```diff
@@ -14,7 +15,8 @@
 /**
  * Represents the link to an external file (e.g. associated PDF file).
  */
-public class LinkedFile {
+//Serialiable is required for drag and drop
```
Comment (koppor) [inline src/main/java/org/jabref/model/entry/LinkedFile.java]: Why not a real class comment? I would always use the class comments.

File: src/main/java/org/jabref/gui/util/ViewModelListCellFactory.java [hunk 2]
```diff
@@ -113,6 +144,21 @@ protected void updateItem(T item, boolean empty) {
                     if (toContextMenu != null) {
                         setContextMenu(toContextMenu.call(viewModel));
                     }
+                    if (toOnDragDetected != null) {
```
Comment (koppor) [inline src/main/java/org/jabref/gui/util/ViewModelListCellFactory.java]: Oh, wow, this code looks odd. Is JavaFX really like that?

**Comment by koppor** [comment]
I have following file entry in my bib entry:

    file      = {:Cla2008 -The Birth of Model Checking.pdf:PDF},

It is rendered like this:

grafik

Does this come from this PR?

**Comment by lenhard** [comment]
I've tested it briefly with an entry with two file links and there reordering was possible and also stored between sessions.

File: src/main/java/org/jabref/gui/util/ViewModelListCellFactory.java [hunk 3]
```diff
@@ -76,6 +82,31 @@
         return this;
     }
 
+    public ViewModelListCellFactory<T> setOnDragDetected(BiConsumer<T, ? super MouseEvent> toOnDragDetected) {
+        this.toOnDragDetected = toOnDragDetected;
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/util/ViewModelListCellFactory.java]: @koppor The null check refers to this factory function where it gets assigned

**Comment by Siedlerchr** [comment]
@koppor  Your display thing is related to the autolinked file. One of your files got autolinked.

File: src/main/java/org/jabref/gui/fieldeditors/LinkedFilesEditor.java [hunk 4]
```diff
@@ -49,12 +56,67 @@ public LinkedFilesEditor(String fieldName, DialogService dialogService, BibDatab
                 .withTooltip(LinkedFileViewModel::getDescription)
                 .withGraphic(LinkedFilesEditor::createFileDisplay)
                 .withContextMenu(this::createContextMenuForFile)
-                .withOnMouseClickedEvent(this::handleItemMouseClick);
+                .withOnMouseClickedEvent(this::handleItemMouseClick)
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/fieldeditors/LinkedFilesEditor.java]: @koppor  The BiConsumer stuff allows to do these fancy things

**Comment by koppor** [comment]
Think, I've  to open another issue. Why the hell is the same file linked
twice? 😱

Am 05.09.2017 20:12 schrieb "Christoph" <notifications@github.com>:

**Pull request merged**